File: p12973.java
import java.util.Stack;

public class p12973 {
    public static int solution(String s)
    {
        Stack<Character> st = new Stack<>();
        int answer = 0;

        for(int i = s.length() - 1; i >= 0; i--){
            if(!st.isEmpty() && st.peek() == s.charAt(i)){
                st.pop();
            }
            else{
                st.push(s.charAt(i));
            }
        }
        if(st.isEmpty()){
            answer = 1;
        }
        return answer;
    }

    public static void main(String[] args){
        System.out.println(solution("cdcd"));
        System.out.println(solution("baabaa"));
    }
}

File: p12973.cpp
#include<bits/stdc++.h>
using namespace std;

int solution(string s) {
    stack<char> st;
    st.push(s[0]);
    int i = 1;
    while(i < s.size()){
        if(!st.empty() && st.top() == s[i]){
            st.pop();
        }
        else{
            st.push(s[i]);
        }
        i++;
    }
    return st.size() == 0 ? 1 : 0;
}

int main(){
    cout << solution("baabaa") << "\n";
    cout << solution("baba") << "\n";
}